File: java/37-sudoku-solver.java
class Solution {
    public void solveSudoku(char[][] board) {
        solve(board, 0);
    }
    private boolean solve(char[][] board, int i) {
        if (i == 81) {
            return true;
        }
        int row = i / 9, col = i % 9;
        if (board[row][col] != '.') {
            return solve(board, i+1);
        }
        boolean[] flag = possibleDigits(board, row, col);
        for (int digit = 1; digit <= 9; digit++) {
            if (flag[digit-1]) {
                board[row][col] = (char)('0'+digit);
                if (solve(board, i+1)) {
                    return true;
                }
            }
        }
        board[row][col] = '.';
        return false;
    }
    private boolean[] possibleDigits(char[][] board, int i, int j) {
        boolean[] flag = new boolean[9];
        Arrays.fill(flag, true);
        for (int index = 0; index < 9; index++) {
            if (board[i][index] != '.') {
                flag[board[i][index] - '1'] = false;
            }
            if (board[index][j] != '.') {
                flag[board[index][j] - '1'] = false;
            }
            int row = i / 3 * 3 + index / 3;
            int col = j / 3 * 3 + index % 3;
            if (board[row][col] != '.') {
                flag[board[row][col] - '1'] = false;
            }
        }
        return flag;
    }
}

File: python/37-sudoku-solver.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        digits = [chr(ord('1') + i) for i in range(9)]

        from collections import defaultdict, deque
        rows = defaultdict(set)
        cols = defaultdict(set)
        secs = defaultdict(set)
        visit = deque([])
        for row in range(9):
            for col in range(9):
                if board[row][col] != ".":
                    rows[row].add(board[row][col])
                    cols[col].add(board[row][col])
                    secs[(row//3,col//3)].add(board[row][col])
                else:
                    visit.append((row,col))

        def solveAttempt():
            if not visit:
                return True
            r, c = visit[0]
            t = (r // 3, c // 3)
            for digit in digits:
                if digit not in rows[r] and digit not in cols[c] and digit not in secs[t]:
                    board[r][c] = digit
                    rows[r].add(digit)
                    cols[c].add(digit)
                    secs[t].add(digit)
                    visit.popleft()
                    if solveAttempt():
                        return True
                    else:
                        board[r][c] = "."
                        rows[r].discard(digit)
                        cols[c].discard(digit)
                        secs[t].discard(digit)
                        visit.appendleft((r,c))
            return False
        
        solveAttempt()
